src/NCPkgFilterService.cc

```diff
@@ -0,0 +1,322 @@
+{
+    bool ret = std::any_of(ZyppRepositoriesBegin(), ZyppRepositoriesEnd(), [](const zypp::Repository& repo) {
+        // if the repository does not belong to any service then the service name is empty
+        return !repo.info().service().empty();
+    });
+
+    yuiMilestone() << "Found a libzypp service: " << ret << std::endl;
+    return ret;
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::fillHeader
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Fill header of servicesitories table (name + URL)
+//
+
+void NCPkgServiceTable::fillHeader()
+{
+    std::vector <std::string> header;
+
+    header.reserve(2);
+    header.push_back( "L" );
+    header.push_back( "L" + NCPkgStrings::PkgName() );
+
+    setHeader( header);
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::addLine
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Add one line (with tag) to the service table
+//
+
+void NCPkgServiceTable::addLine ( ZyppService r, const std::vector <std::string> & cols )
+{
+    //use default ctor, add cell in the next step
+    YTableItem *tabItem = new YTableItem();
+
+    //place tag (with service reference) to the 0th column
+    tabItem->addCell( new NCPkgServiceTag ( r ) );
+
+    // and append the rest (name, URL and stuff)
+    for ( unsigned i = 1; i < cols.size() + 1; ++i ) {
+	tabItem->addCell( cols[ i-1 ]);
+    }
+
+    // this is NCTable::addItem( tabItem );
+    //it actually appends the line to the table
+    addItem( tabItem );
+
+
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::getTag
+//	METHOD TYPE : NCPkgServiceTag *
+//
+//	DESCRIPTION : Get tag of service table line on current index,
+//		      ( contains service reference)
+//
+
+NCPkgServiceTag* NCPkgServiceTable::getTag (const int & index )
+{
+   NCTableLine *line = myPad()->ModifyLine( index );
+   if ( !line )
+   {
+	return 0;
+   }
+
+   YTableItem *it = dynamic_cast<YTableItem*> (line->origItem() );
+
+   //get actual service tag from 0th column of the table
+   YTableCell *tcell = it->cell(0);
+   NCPkgServiceTag *tag = static_cast<NCPkgServiceTag *>( tcell );
+
+   return tag;
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::getService
+//	METHOD TYPE : ZyppService
+//
+//	DESCRIPTION : Get service reference from selected line's tag
+//
+
+ZyppService NCPkgServiceTable::getService( int index )
+{
+    NCPkgServiceTag *t = getTag( index );
+
+    if ( !t )
+    {
+	return ZyppService( );
+    }
+    else
+    {
+	 return t->getService();
+    }
+}
+
+std::string NCPkgServiceTable::showDescription( ZyppService r)
+{
+    zypp::ServiceInfo si = repo_manager->getService(r);
+
+    std::string label = _( "<b>Service URL:</b>" );
+    std::string ret = label + si.url().asString();
```

Comment:
It's unlikely that the URL will contain any special HTML character (<>&") but in general it should be HTML-escaped before adding to rich text.

Reply:
Actually it is quite *likely* that a URL will contain a `&`. Our usual zypp URLs don't but I have just tested with a handcrafted service `url = http://example.com/blah?foo=bar&baz=qux` and YAY, even got a segfault :)

Reply:
(The segfault was unrelated, caused by my misguided attempt to introduce `shared_ptr` in the middle of a legacy API)